README.md

```diff
@@ -1,2 +1,80 @@
-# beacon-starter
+# Beacon Reader Example
 
+> A starter project for reading beacon values from a smart contract
+
+This project is composed of the following:
+
+1. A test that uses a mocked RrpBeaconServer to simulate reading a beacon value from a smart contract.
+1. A script to deploy the BeaconReaderExample contract to any network. This script will set the address of the RrpBeaconServer contract deployed on the selected network and after the script finishes it will display the address of the contract deployed but also save this address to `deployments` directory on the root of this repo (except for hardhat network since this network is started and stop just for the duration of the script execution).
+1. A script to read the beacon value from the deployed contract. If selected network is not `hardhat` or `localhost` then this address must be then whitelisted by the RrpBeaconServer manager account prior to reading the beacon value.
+
+## Instructions
+
+1. Clone this repo
+1. Run the following to install the dependencies
+
+```sh
+npm install
+```
+
+Now depending on the network you are using you will need to do the following:
+
+### Default network
+
+Simplest way to test this project is by using the integrated [Hardhat network](https://hardhat.org/hardhat-network/) and not extra steps are needed in this case to run both the test and the scripts.
+
+#### Test
+
+```sh
+npx hardhat test test/BeaconReaderExample.test.js
+```
+
+#### Script
+
+```sh
+npx hardhat run scripts/deploy.js
+npx hardhat run scripts/read-beacon.js
+```
+
+### Localhost network
+
+You could also start a local ethreum node by running the following command on a separate terminal:
+
+```sh
+hardhat node
+```
+
+#### Test
+
+```sh
+npx hardhat test test/BeaconReaderExample.test.js --network localhost
+```
+
+#### Script
+
+```sh
+npx hardhat run scripts/deploy.js --network localhost
+npx hardhat run scripts/read-beacon.js --network localhost
+```
+
+### Remote networks
+
+This will require that you set some parameters in an .env file. You could rename the [.env.example](./.env.example) file in the root of this repo and replace the placeholders with valid values.
+
+For instance if you wanted to deploy the contract on [Polygon Mumbai Testnet](https://docs.polygon.technology/docs/develop/network-details/network/) you will need to set `NETWORK=polygon-mumbai` and `PROVIDER_URL=https://rpc-mumbai.matic.today`. Keep in mind that you'll need to also set the `MNEMONIC` of an account that needs to have enough funds.
+
+#### Test
+
+```sh
+npx hardhat test test/BeaconReaderExample.test.js --network polygon-mumbai
```

Comment:
This succeeded when I used `rinkeby`, but on `polygon-mumbai` I ran into the cryptic error below. This is with a freshly funded wallet containing 0.5 MATIC and it was the same error for both infura and the `PROVIDER_URL` above.
```sh
> npx hardhat test test/BeaconReaderExample.test.js --network polygon-mumbai


  BeaconReaderExample
    1) Should return the beacon value


  0 passing (21s)
  1 failing

  1) BeaconReaderExample
       Should return the beacon value:
     Error: call revert exception (method="readBeacon(bytes32)", errorArgs=null, errorName=null, errorSignature=null, reason=null, code=CALL_EXC
EPTION, version=abi/5.5.0)
      at Logger.makeError (node_modules/@ethersproject/logger/src.ts/index.ts:225:28)
      at Logger.throwError (node_modules/@ethersproject/logger/src.ts/index.ts:237:20)
      at Interface.decodeFunctionResult (node_modules/@ethersproject/abi/src.ts/interface.ts:425:23)
      at Contract.<anonymous> (node_modules/@ethersproject/contracts/src.ts/index.ts:392:44)
      at step (node_modules/@ethersproject/contracts/lib/index.js:48:23)
      at Object.next (node_modules/@ethersproject/contracts/lib/index.js:29:53)
      at fulfilled (node_modules/@ethersproject/contracts/lib/index.js:20:58)
      at processTicksAndRejections (internal/process/task_queues.js:95:5)
```

Reply:
I'm getting the same on polygon and interestingly enough the address of the supposed BeaconReaderExample contract displayed by the deploy script, doesn't seem to be correct when going and checking it on https://mumbai.polygonscan.com/ 🤔

Reply:
Well, it seems to be an issue between hardhat and mumbai testnet https://github.com/nomiclabs/hardhat/issues/2162

Reply:
Workaround of waiting for the contract creation tx to complete seems to do the trick.